include/oneapi/dpl/pstl/hetero/memory_ranges_impl_hetero.h

```diff
@@ -0,0 +1,246 @@
+
+    auto __last = std::ranges::begin(__r) + std::ranges::size(__r);
+
+    if constexpr (!std::is_trivially_default_constructible_v<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__op_uninitialized_default_construct<std::decay_t<_ExecutionPolicy>>{},
+            oneapi::dpl::__ranges::views::all(std::forward<_R>(__r)));
+    }
+
+    return __last;
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_uninitialized_value_construct
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_uninitialized_value_construct(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r)
+{
+    using _ValueType = std::ranges::range_value_t<_R>;
+
+    auto __last = std::ranges::begin(__r) + std::ranges::size(__r);
+
+    if constexpr (oneapi::dpl::__internal::__trivial_uninitialized_value_construct<_ValueType>)
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__brick_fill<__hetero_tag<_BackendTag>, _ValueType>{_ValueType()},
+            std::forward<_R>(__r));
+    }
+    else
+    {
+        oneapi::dpl::__internal::__ranges::__pattern_walk_n(
+            __tag, std::forward<_ExecutionPolicy>(__exec),
+            oneapi::dpl::__internal::__op_uninitialized_value_construct<std::decay_t<_ExecutionPolicy>>{},
+            std::forward<_R>(__r));
+    }
+
+    return __last;
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_uninitialized_copy
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _InRange, typename _OutRange>
+std::ranges::uninitialized_copy_result<std::ranges::borrowed_iterator_t<_InRange>,
+                                       std::ranges::borrowed_iterator_t<_OutRange>>
+__pattern_uninitialized_copy(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r,
+                             _OutRange&& __out_r)
+{
+    using _OutValueType = std::ranges::range_value_t<_OutRange>;
+    using _OutRefType = std::ranges::range_reference_t<_OutRange>;
+    using _InRefType = std::ranges::range_reference_t<_InRange>;
+
+    using _Size = std::common_type_t<std::ranges::range_size_t<_InRange>, std::ranges::range_size_t<_OutRange>>;
+    const _Size __n1 = std::ranges::size(__in_r);
+    const _Size __n2 = std::ranges::size(__out_r);
+    const _Size __n = std::min(__n1, __n2);
```

Comment:
`__n1` and `__n2` do not seem to be used beyond computing `__n`; maybe eliminate?
```suggestion
    const _Size __n = std::min<_Size>(std::ranges::size(__in_r), std::ranges::size(__out_r));
```

Same for `uninitialized_move`.